### PR TITLE
Add section on nonlinear equations

### PR DESCRIPTION
**Pull request by manuhuth**
Implement section on nonlinear equations. 

- add source to bib file
- add nonlinear_equations.py
- add tests
- add rst documentation file

### CODE REVIEW

**Comment by manuhuth** [comment]
The Read the docs batch still does not appear. I have checked the trobleshootong section from [the documentation](https://docs.readthedocs.io/en/stable/guides/autobuild-docs-for-pull-requests.html) but both points are met. 
Do we have enabled the relevant section in read the docs?

**Comment by amageh** [comment]
> The Read the docs batch still does not appear. I have checked the trobleshootong section from [the documentation](https://docs.readthedocs.io/en/stable/guides/autobuild-docs-for-pull-requests.html) but both points are met.
> Do we have enabled the relevant section in read the docs?

I found the issue, it should be working now :). You'll need to add `nonlinear_equations` to the toc in `docs/source/index` for it to appear in the documentation though.

**Comment by manuhuth** [comment]
I have uploaded the respective Files:

- add source to bib file
- add nonlinear_equations.py
- add tests
- add rst documentation file

additionally I have fixed some displaying bugs in 'optimization.py'

**Review by amageh — CHANGES_REQUESTED** [review]
Nice work and good job on the docstrings! I left some comments. Most of them are just spaces or typos, you can just commit those in a batch directly if you want. 

Some other points:
- It is generally discouraged to use mutable objects as defaults in functions ([this](https://florimond.dev/blog/articles/2018/08/python-mutable-defaults-are-the-source-of-all-evil/) is an entertaining and short read on why) i.e. PEP8 is complaining about all the lists that are currently used as default arguments in the functions. The article suggests assigning the default inside the function instead, so I would suggest doing that.
- For function inputs that need to have a certain form or length, you can maybe add an assert statement or exception to test for this and raise an error if it's not fulfilled (this makes the functions more user friendly but it obviously is more work so I am just suggesting it here).
- This is minor and very optional (for the future), but I would prefer it if you wrote more meaningful commit messages (at least for larger/important commits), since scanning those in the beginning of the review gives a nice overview of the changes that have been made/ things which have been added). This doesn't matter for the content though, just makes life easier for the reviewer :).

File: temfpy/nonlinear_equations.py [hunk 1]
```diff
@@ -0,0 +1,987 @@
...
+        The default value is 1.
+
+    Returns
+    -------
+    array_like
+        Output domain
+    numpy.array
+        Analytically derived Jacobian
+    numpy.array
+        Numerically derived Jacobian
```
[the remark this answers is not in the document]
Comment (manuhuth) [reply temfpy/nonlinear_equations.py]: I chose to write it as array_like and then indicate the tuple :)

File: temfpy/nonlinear_equations.py [hunk 2]
```diff
@@ -0,0 +1,987 @@
...
+    rslt = (
+        (np.exp(x) + x_im1 - b)
+        * np.concatenate((a, np.array(range(2, p + 1))), axis=None)
+        / a
+    )
+
+    return np.array(rslt)
+
+
+def _exponential_jacobian(x, a=10, b=1):
```
[the remark this answers is not in the document]
Comment (manuhuth) [reply temfpy/nonlinear_equations.py:129]: True. Fully erased by the derivative.

File: temfpy/nonlinear_equations.py [hunk 3]
```diff
@@ -0,0 +1,987 @@
...
+        x = np.array(x)
+        y = np.array(y)
+
+        term_sum = []
+        for i in range(0, p):
+            term_sum.append(np.sum(x / (y[i] + y)))
+
+        rslt = x - 1 / (1 - c * y / (2 * p) * term_sum)
+
+        return np.array(rslt)
```
[the remark this answers is not in the document]
Comment (manuhuth) [reply temfpy/nonlinear_equations.py:1046]: My problem was that for the chandrasekahr equation x, y and c are parameters that must be chosen. However, the vector of interest is only x. Hence, the numerical Jacobian must be evaluated of a function that only depends on x (fixed values cannot be given to the jacobian function of numdifftools - to my google knowledge). Therefore this seemed to be a reasonable approach to me. 

However, I am open for other suggestions! :)

[the remark this answers is not in the document]
Comment (amageh) [reply temfpy/nonlinear_equations.py:1046]: Ah I see! For this [functools.partial](https://docs.python.org/2/library/functools.html) may be an elegant solution :-)

**Review by amageh — APPROVED** [review]

**Pull request merged**